File: python/backtest.py
"""
Backtesting framework for saliency-based trading strategies.

This module provides tools for backtesting trading strategies that use
saliency maps for signal filtering and position sizing.
"""

import torch
import torch.nn as nn
import numpy as np
import pandas as pd
from typing import Optional, List, Dict, Tuple, Callable
from dataclasses import dataclass
from .saliency import SaliencyComputer, compute_saliency_concentration, aggregate_feature_importance


@dataclass
class Trade:
    """Represents a single trade."""
    entry_date: pd.Timestamp
    exit_date: pd.Timestamp
    direction: int
    entry_price: float
    exit_price: float
    position_size: float
    pnl: float
    pnl_pct: float


@dataclass
class BacktestResult:
    """Results from a backtest."""
    total_return: float
    sharpe_ratio: float
    sortino_ratio: float
    max_drawdown: float
    win_rate: float
    profit_factor: float
    num_trades: int
    trades: List[Trade]
    equity_curve: pd.Series
    daily_returns: pd.Series


class SaliencyTradingStrategy:
    """
    Trading strategy that uses saliency maps for filtering and sizing.

    Args:
        model: Trained PyTorch model
        saliency_method: Method for computing saliency ('vanilla', 'gradient_x_input', etc.)
        min_confidence: Minimum prediction confidence to trade
        min_concentration: Minimum saliency concentration to trade
        interpretable_features: Indices of features that should be important for valid signals
        position_sizing: 'fixed', 'concentration', or 'confidence'
    """

    def __init__(
        self,
        model: nn.Module,
        saliency_method: str = 'gradient_x_input',
        min_confidence: float = 0.6,
        min_concentration: float = 0.3,
        interpretable_features: Optional[List[int]] = None,
        position_sizing: str = 'fixed'
    ):
        self.model = model
        self.model.eval()
        self.saliency_computer = SaliencyComputer(model)
        self.saliency_method = saliency_method
        self.min_confidence = min_confidence
        self.min_concentration = min_concentration
        self.interpretable_features = interpretable_features
        self.position_sizing = position_sizing

    def compute_saliency(self, x: torch.Tensor) -> torch.Tensor:
        """Compute saliency using the configured method."""
        if self.saliency_method == 'vanilla':
            return self.saliency_computer.vanilla_gradient(x)
        elif self.saliency_method == 'gradient_x_input':
            return self.saliency_computer.gradient_x_input(x)
        elif self.saliency_method == 'integrated':
            return self.saliency_computer.integrated_gradients(x)
        elif self.saliency_method == 'smoothgrad':
            return self.saliency_computer.smoothgrad(x)
        else:
            raise ValueError(f"Unknown saliency method: {self.saliency_method}")

    def check_interpretability(
        self,
        saliency: torch.Tensor,
        top_k: int = 3
    ) -> bool:
        """
        Check if the top salient features are interpretable.

        Args:
            saliency: Saliency map
            top_k: Number of top features to check

        Returns:
            True if top features are in the interpretable set
        """
        if self.interpretable_features is None:
            return True

        feature_importance = aggregate_feature_importance(saliency)
        if feature_importance.dim() == 2:
            feature_importance = feature_importance[0]

        top_features = feature_importance.topk(top_k).indices.tolist()

        overlap = len(set(top_features) & set(self.interpretable_features))
        return overlap >= (top_k // 2 + 1)

    def generate_signal(
        self,
        x: torch.Tensor
    ) -> Tuple[int, float]:
        """
        Generate a trading signal.

        Args:
            x: Input tensor of shape (1, sequence_length, features)

        Returns:
            Tuple of (signal, position_size) where signal is -1, 0, or 1
        """
        with torch.no_grad():
            prediction = self.model(x).item()

        saliency = self.compute_saliency(x)
        concentration = compute_saliency_concentration(saliency).item()

        if not self.check_interpretability(saliency):
            return 0, 0.0

        if concentration < self.min_concentration:
            return 0, 0.0

        if prediction > self.min_confidence:
            signal = 1
        elif prediction < (1 - self.min_confidence):
            signal = -1
        else:
            return 0, 0.0

        if self.position_sizing == 'fixed':
            position_size = 1.0
        elif self.position_sizing == 'concentration':
            position_size = concentration
        elif self.position_sizing == 'confidence':
            position_size = abs(prediction - 0.5) * 2
        else:
            position_size = 1.0

        return signal, position_size


def run_backtest(
    strategy: SaliencyTradingStrategy,
    features: np.ndarray,
    prices: pd.Series,
    dates: pd.DatetimeIndex,
    initial_capital: float = 100000,
    transaction_cost: float = 0.001,
    max_position: float = 1.0
) -> BacktestResult:
    """
    Run a backtest on the given data.

    Args:
        strategy: Trading strategy to test
        features: Feature array of shape (num_samples, sequence_length, num_features)
        prices: Price series
        dates: Date index
        initial_capital: Starting capital
        transaction_cost: Transaction cost as fraction of trade value
        max_position: Maximum position size as fraction of capital

    Returns:
        BacktestResult object
    """
    capital = initial_capital
    position = 0
    position_size = 0.0
    entry_price = 0.0
    entry_date = None

    trades: List[Trade] = []
    equity = [capital]
    equity_dates = [dates[0]]

    for i in range(len(features)):
        x = torch.FloatTensor(features[i:i+1])
        current_price = prices.iloc[i]
        current_date = dates[i]

        signal, size = strategy.generate_signal(x)

        if position != 0 and signal != position:
            exit_price = current_price
            trade_cost = abs(position_size * exit_price * transaction_cost)

            if position == 1:
                pnl = position_size * (exit_price - entry_price) - trade_cost
            else:
                pnl = position_size * (entry_price - exit_price) - trade_cost

            pnl_pct = pnl / (position_size * entry_price)

            capital += pnl

            trades.append(Trade(
                entry_date=entry_date,
                exit_date=current_date,
                direction=position,
                entry_price=entry_price,
                exit_price=exit_price,
                position_size=position_size,
                pnl=pnl,
                pnl_pct=pnl_pct
            ))

            position = 0
            position_size = 0.0

        if signal != 0 and position == 0:
            position = signal
            entry_price = current_price
            entry_date = current_date
            position_size = size * max_position * capital / current_price
            trade_cost = position_size * current_price * transaction_cost
            capital -= trade_cost

        if position != 0:
            if position == 1:
                current_equity = capital + position_size * (current_price - entry_price)
            else:
                current_equity = capital + position_size * (entry_price - current_price)
            equity.append(current_equity)
        else:
            equity.append(capital)

        equity_dates.append(current_date)

    equity_curve = pd.Series(equity, index=equity_dates)
    daily_returns = equity_curve.pct_change().dropna()

    total_return = (equity_curve.iloc[-1] / initial_capital) - 1
    sharpe_ratio = compute_sharpe_ratio(daily_returns)
    sortino_ratio = compute_sortino_ratio(daily_returns)
    max_drawdown = compute_max_drawdown(equity_curve)

    if len(trades) > 0:
        win_rate = sum(1 for t in trades if t.pnl > 0) / len(trades)
        gross_profit = sum(t.pnl for t in trades if t.pnl > 0)
        gross_loss = abs(sum(t.pnl for t in trades if t.pnl < 0))
        profit_factor = gross_profit / (gross_loss + 1e-10)
    else:
        win_rate = 0.0
        profit_factor = 0.0

    return BacktestResult(
        total_return=total_return,
        sharpe_ratio=sharpe_ratio,
        sortino_ratio=sortino_ratio,
        max_drawdown=max_drawdown,
        win_rate=win_rate,
        profit_factor=profit_factor,
        num_trades=len(trades),
        trades=trades,
        equity_curve=equity_curve,
        daily_returns=daily_returns
    )


def compute_sharpe_ratio(
    returns: pd.Series,
    risk_free_rate: float = 0.0,
    annualization_factor: float = 252
) -> float:
    """Compute annualized Sharpe ratio."""
    excess_returns = returns - risk_free_rate / annualization_factor
    if excess_returns.std() == 0:
        return 0.0
    return np.sqrt(annualization_factor) * excess_returns.mean() / excess_returns.std()


def compute_sortino_ratio(
    returns: pd.Series,
    risk_free_rate: float = 0.0,
    annualization_factor: float = 252
) -> float:
    """Compute annualized Sortino ratio."""
    excess_returns = returns - risk_free_rate / annualization_factor
    downside_returns = excess_returns[excess_returns < 0]
    if len(downside_returns) == 0 or downside_returns.std() == 0:
        return 0.0
    downside_std = downside_returns.std()
    return np.sqrt(annualization_factor) * excess_returns.mean() / downside_std


def compute_max_drawdown(equity_curve: pd.Series) -> float:
    """Compute maximum drawdown."""
    rolling_max = equity_curve.expanding().max()
    drawdowns = (equity_curve - rolling_max) / rolling_max
    return abs(drawdowns.min())


def compare_strategies(
    strategies: Dict[str, SaliencyTradingStrategy],
    features: np.ndarray,
    prices: pd.Series,
    dates: pd.DatetimeIndex,
    **kwargs
) -> pd.DataFrame:
    """
    Compare multiple strategies.

    Args:
        strategies: Dictionary of strategy name to strategy object
        features: Feature array
        prices: Price series
        dates: Date index
        **kwargs: Additional arguments for run_backtest

    Returns:
        DataFrame with comparison metrics
    """
    results = {}

    for name, strategy in strategies.items():
        result = run_backtest(strategy, features, prices, dates, **kwargs)
        results[name] = {
            'Total Return': f'{result.total_return:.2%}',
            'Sharpe Ratio': f'{result.sharpe_ratio:.2f}',
            'Sortino Ratio': f'{result.sortino_ratio:.2f}',
            'Max Drawdown': f'{result.max_drawdown:.2%}',
            'Win Rate': f'{result.win_rate:.2%}',
            'Profit Factor': f'{result.profit_factor:.2f}',
            'Num Trades': result.num_trades
        }

    return pd.DataFrame(results).T

File: rust_examples/examples/backtest.rs
//! Example: Backtesting saliency-based trading strategies
//!
//! This example demonstrates how to backtest trading strategies
//! that use saliency maps for signal filtering.

use ndarray::Array2;
use saliency_maps_trading::{
    models::network::TradingNetwork,
    saliency::gradient::SaliencyTrader,
};

/// Simple backtest result
struct BacktestResult {
    total_return: f64,
    sharpe_ratio: f64,
    max_drawdown: f64,
    num_trades: usize,
    win_rate: f64,
}

fn main() {
    env_logger::init();

    println!("Saliency Maps Trading - Backtest");
    println!("=================================\n");

    // Configuration
    let sequence_length = 30;
    let num_features = 5;
    let num_samples = 200;

    // Generate synthetic price data
    println!("Generating synthetic market data...");
    let (samples, prices) = generate_synthetic_data(num_samples, sequence_length, num_features);
    println!("  Generated {} samples", samples.len());

    // Test different strategies
    println!("\n{:-<70}", "");
    println!("Strategy Comparison:");
    println!("{:-<70}\n", "");

    // Strategy 1: Baseline (no saliency filtering)
    let baseline_network = TradingNetwork::new(sequence_length, num_features, &[64, 32]);
    let baseline_result = run_backtest(
        &baseline_network,
        &samples,
        &prices,
        false, // no saliency filter
        0.0,   // no concentration threshold
    );
    print_result("Baseline (No Filter)", &baseline_result);

    // Strategy 2: Saliency concentration filter
    let saliency_network = TradingNetwork::new(sequence_length, num_features, &[64, 32]);
    let saliency_result = run_backtest(
        &saliency_network,
        &samples,
        &prices,
        true,  // use saliency filter
        0.3,   // concentration threshold
    );
    print_result("Saliency Filtered", &saliency_result);

    // Strategy 3: Higher concentration threshold
    let strict_network = TradingNetwork::new(sequence_length, num_features, &[64, 32]);
    let strict_result = run_backtest(
        &strict_network,
        &samples,
        &prices,
        true,
        0.5,  // stricter concentration
    );
    print_result("Strict Saliency", &strict_result);

    // Summary comparison
    println!("\n{:-<70}", "");
    println!("Summary:");
    println!("{:-<70}\n", "");

    println!(
        "{:<20} {:>12} {:>12} {:>12} {:>10} {:>10}",
        "Strategy", "Return", "Sharpe", "Max DD", "Trades", "Win Rate"
    );
    println!("{:-<70}", "");

    for (name, result) in [
        ("Baseline", &baseline_result),
        ("Saliency", &saliency_result),
        ("Strict", &strict_result),
    ] {
        println!(
            "{:<20} {:>11.2}% {:>12.2} {:>11.2}% {:>10} {:>9.1}%",
            name,
            result.total_return * 100.0,
            result.sharpe_ratio,
            result.max_drawdown * 100.0,
            result.num_trades,
            result.win_rate * 100.0
        );
    }

    println!("\n✓ Backtest complete!");
}

/// Generate synthetic market data for testing
fn generate_synthetic_data(
    num_samples: usize,
    sequence_length: usize,
    num_features: usize,
) -> (Vec<Array2<f64>>, Vec<f64>) {
    let mut samples = Vec::new();
    let mut prices = Vec::new();

    let mut price = 100.0;

    for _ in 0..num_samples {
        // Generate sequence
        let mut sequence = Array2::zeros((sequence_length, num_features));

        for t in 0..sequence_length {
            // Random walk with mean reversion
            let change = (rand::random::<f64>() - 0.5) * 2.0 - (price - 100.0) * 0.01;
            price = (price + change).max(50.0).min(150.0);

            sequence[[t, 0]] = price - 0.5 + rand::random::<f64>(); // Open
            sequence[[t, 1]] = price + rand::random::<f64>() * 2.0;  // High
            sequence[[t, 2]] = price - rand::random::<f64>() * 2.0;  // Low
            sequence[[t, 3]] = price;  // Close
            sequence[[t, 4]] = 1000.0 + rand::random::<f64>() * 500.0;  // Volume
        }

        // Normalize
        for j in 0..num_features {
            let col_mean: f64 = (0..sequence_length)
                .map(|i| sequence[[i, j]])
                .sum::<f64>() / sequence_length as f64;
            let col_std: f64 = {
                let variance: f64 = (0..sequence_length)
                    .map(|i| (sequence[[i, j]] - col_mean).powi(2))
                    .sum::<f64>() / sequence_length as f64;
                variance.sqrt().max(1e-10)
            };

            for i in 0..sequence_length {
                sequence[[i, j]] = (sequence[[i, j]] - col_mean) / col_std;
            }
        }

        samples.push(sequence);
        prices.push(price);
    }

    (samples, prices)
}

/// Run a simple backtest
fn run_backtest(
    network: &TradingNetwork,
    samples: &[Array2<f64>],
    prices: &[f64],
    use_saliency: bool,
    concentration_threshold: f64,
) -> BacktestResult {
    let mut equity = 100000.0;
    let initial_equity = equity;
    let mut position = 0;  // 1 = long, -1 = short, 0 = flat
    let mut entry_price = 0.0;

    let mut trades: Vec<f64> = Vec::new();
    let mut equity_history = vec![equity];
    let mut max_equity = equity;
    let mut max_drawdown = 0.0;

    let trader = if use_saliency {
        Some(SaliencyTrader::new(
            network.clone(),
            0.55,
            concentration_threshold,
            vec![3, 4],  // Close, Volume
        ))
    } else {
        None
    };

    for i in 0..samples.len().saturating_sub(1) {
        let current_price = prices[i];
        let next_price = prices[i + 1];

        // Generate signal
        let (signal, _size) = if let Some(ref t) = trader {
            t.generate_signal(&samples[i])
        } else {
            // Simple prediction without filtering
            let pred = network.forward(&samples[i]);
            if pred > 0.55 {
                (1, 1.0)
            } else if pred < 0.45 {
                (-1, 1.0)
            } else {
                (0, 0.0)
            }
        };

        // Close existing position if signal changes
        if position != 0 && signal != position {
            let pnl = if position == 1 {
                (current_price - entry_price) / entry_price
            } else {
                (entry_price - current_price) / entry_price
            };
            equity *= 1.0 + pnl;
            trades.push(pnl);
            position = 0;
        }

        // Open new position
        if position == 0 && signal != 0 {
            position = signal;
            entry_price = current_price;
        }

        // Update equity tracking
        equity_history.push(equity);
        if equity > max_equity {
            max_equity = equity;
        }
        let drawdown = (max_equity - equity) / max_equity;
        if drawdown > max_drawdown {
            max_drawdown = drawdown;
        }
    }

    // Close final position
    if position != 0 && !prices.is_empty() {
        let final_price = prices[prices.len() - 1];
        let pnl = if position == 1 {
            (final_price - entry_price) / entry_price
        } else {
            (entry_price - final_price) / entry_price
        };
        equity *= 1.0 + pnl;
        trades.push(pnl);
    }

    // Calculate metrics
    let total_return = (equity - initial_equity) / initial_equity;

    let win_rate = if trades.is_empty() {
        0.0
    } else {
        trades.iter().filter(|&&t| t > 0.0).count() as f64 / trades.len() as f64
    };

    let sharpe_ratio = if trades.len() > 1 {
        let mean: f64 = trades.iter().sum::<f64>() / trades.len() as f64;
        let variance: f64 = trades.iter().map(|t| (t - mean).powi(2)).sum::<f64>()
            / trades.len() as f64;
        let std = variance.sqrt();
        if std > 1e-10 {
            (mean / std) * (252.0_f64).sqrt()  // Annualized
        } else {
            0.0
        }
    } else {
        0.0
    };

    BacktestResult {
        total_return,
        sharpe_ratio,
        max_drawdown,
        num_trades: trades.len(),
        win_rate,
    }
}

fn print_result(name: &str, result: &BacktestResult) {
    println!("{}", name);
    println!("  Total Return: {:.2}%", result.total_return * 100.0);
    println!("  Sharpe Ratio: {:.2}", result.sharpe_ratio);
    println!("  Max Drawdown: {:.2}%", result.max_drawdown * 100.0);
    println!("  Number of Trades: {}", result.num_trades);
    println!("  Win Rate: {:.1}%\n", result.win_rate * 100.0);
}
